### ISSUE TITLE
Required API for slice()

### ISSUE THREAD

**Issue by MarcusKlik** [krlmlr/fstplyr#4]
Method `fst::read_fst` needs an API change to accommodate `slice()`. In addition to parameters `from` and `to` a  (possibly randomly ordered) `integer` vector should be accepted.

The [current implementation](https://github.com/krlmlr/fstplyr/blob/master/R/tbl.R#L50-L53) for `read_from_meta` reads the values between the `min` and `max` row number, which possibly reads many rows from file that are not requested.

This won't be the only API change needed, perhaps it's best to define a new method in `fst` which is tailored for `fstplyr` requests?

**Comment by krlmlr** [comment]
A two-stage approach may be a good way here:

1. Transform a randomly ordered `integer` vector into an opaque index data structure that is optimized for reading
    - Need to account for negative indexes and out-of-bound, too
2. `read_fst()` accepts either an integer vector (to which it applies step 1) or an already processed index d.s.

That opaque data structure would contain a vector of rows to be read, and a map into that vector. Can we isolate the code for step 1, so that it can be later moved/copied to another package (tidyselect or dplyr)?

**Comment by MarcusKlik** [comment]
Yes, having the map would speed up subsequent operations and the sorting would have to be done only once.

Also, the map could be reduced in size when we `slice()` less than 1/16 of the total number of rows. If that is the case, the map could be replaced with a binary vector containing a single bit per row (sort of a _row mask_). With such a mask, reading could be done even faster (in C++). The mask would need an `R` container such as a `raw` vector to hold it.

Would the slice field in the `tibble` hold the index data-structure as well?

**Comment by krlmlr** [comment]
Thanks. Yes, the `tbl_fst` structure should then hold the index d.s., which is computed only once.

The most challenging part will be to combine indexes for sequences of `slice()` calls, as in `x[i1][i2]` -- it was more difficult than anticipated for the column names, even with the help of tidyselect. Because of this, I'd rather focus on implementing just one kind of data structure.

I'm curious, though -- why is parsing a bitmap faster than iterating a sequence of pointers? I'm not sure we're talking about the same mapping data structure. For `slice(data, i)` I was thinking about:

- a vector with the same length as `unique(i)` that contains the rows to be fetched in sorted order
- a vector with the same length as `i` that contains a pointer into the first vector for each row to be fetched

For efficient reading we'd also want to store the reverse map, maybe using two more vectors with the same dimensions which hold an array of linked lists (to support `slice(data, c(1, 1))`)? We can omit the second vector if the slice doesn't contain duplicate rows (which is true after most operations including `filter()` and `arrange()`).

**Comment by MarcusKlik** [comment]
Yes, great, that should hold all the information needed for fast retrieval.
With a bitmap you can iterate through the vector in blocks of 8, 16, 32 or 64 rows and check if all bits equal zero with a single statement  (just check if the `int8` or `int16`, etc. value equals 0). For sparse selections, that will speed up the selection significantly I think. There are also some SIMD instructions that specialize in selecting from a vector using a bitmap, but I don't think SIMD is being used in any `R` package yet.

Especially the 1:n slice sequences are very expensive in terms of memory and it would require an `if` statement for each element to determine if it's still inside the data block that has been read from disk (or some binary search).

But that's just intuition, I'll have to benchmark to know for sure!

**Comment by krlmlr** [comment]
I need a better understanding of fst internals to fully follow this argument, I'd like to leave that to you for now ;-) We just need to keep in mind that the index d.s. must support duplicates (`slice(data, c(1, 1))`) and chaining (`data %>% slice(i1) %>% slice(i2)`).
